### ISSUE TITLE
Secure the password hash better

### ISSUE THREAD

**Issue by xZise** [espinosaluis/TrackMeViewer#2]
At the moment the password is salted and hashed which is at least something, but for one the salt is static so it does not hash the same password differently (so if multiple users are using the same password it creates the same hash, one issue that salting it tries to solve it).

Now PHP supports since 5.5 the `password_hash` function which is using bcrypt and there is a backport to 5.3 (or even older) at https://github.com/ircmaxell/password_compat.

I don't know which versions you want to support but the one server I'm currently using it on is using PHP 5.3 so I'd prefer that it includes the backport.

It does require that the SQL queries are rewritten, as you cannot search for the password anymore, but you need to verify the password. Now I've been working on switching to PDO (to preven SQL injections) and so maybe I could lay the groundwork (with a function like “validLogin(username, password)”) and then a future patch could replace the static salt with `password_hash` and `password_verify`.

**Comment by espinosaluis** [comment]
xZise,

At this moment I cannot change anything on the client (I'm working on other projects), so I'm ok with any changes that you consider about that but if it doesn't require any changes on the client side.
Otherwise, we may talk about this again in the future.

**Comment by xZise** [comment]
Does the client do direct queries to the database? If it's via the scripts here we coule modify the requests so that it appears to the outside that nothing has changed.

**Comment by espinosaluis** [comment]
No, that's not possible because most public web hostings don't allow that. That's the reason I did all communication using HTTP calls using the PHP pages.

**Comment by xZise** [comment]
So in that case we could look through the code and change the related code. I actually have a future patch which makes it easier to prepare for this. I just need to get around to improve all files (as I mentioned in the opening post). That system is already implemented in such a way that it is easy to support `password_hash` and `password_verify`. The only major part missing then is to detect which type it's using and optionally upgrade (there is `password_needs_rehash` or something similar which could be integrated into it).

From the outside the password is not hashed, so clients accessing it from the outside shouldn't be affected. The only issue that could arise is when cookies are used, because he cookie is storing the password hashed. But that cookie should use a session id anyway (I'll upload a pull request soon which is using PHP's `$_SESSION` constant).

**Comment by espinosaluis** [comment]
Sounds good. If you require changes on the client side, I could do that but only if they are minimal at this time. As I mentioned, I'm right now working on other projects so I can't really make any modifications. Thanks for your help.

**Comment by espinosaluis** [comment]
xZise, I found a problem. When selecting a route, it takes me back to the login screen. I'm using Chrome.

**Comment by xZise** [comment]
I think it is missing a `session_start()` call at the start of the document.

**Comment by xZise** [comment]
Okay the issue with the patch that introduced `password_verify` and `password_hash` is that if your PHP version is new enough to exist, it won't work because it doesn't allow to redefine the two methods.

Not sure what the best approach is here. Because to use PHP's functions we'd need to update all places (especially `requests.php` which I haven't tested at all yet) and we'd need some way to support older PHP installations.
